addons/udes_stock_routing/models/stock_move.py

```diff
@@ -105,15 +106,163 @@ def _get_new_picking_values(self):
+
+        return res
+
     def _action_assign(self):
         """
-        Extend action_assign to include the two stage hook.
+        Extend action_assign to include the two stage hook and split pick hook.
         This is done at the move level incase refactoring splits picks after assigning stock.
         This way, we don't lose any moves which need to be split to 2 stage but got refactored out.
         """
-        res = super()._action_assign()
+        StockMoveObj = self.env["stock.move"]
+        Rule = self.env["stock.rule"]
+
+        res = StockMoveObj.browse()
+        applicable_rules = Rule.browse()
+        # Allows us to build data on which moves have been split for which rules
+        # which we can then split to backorder pickings after looping over all rules.
+        moves_to_split_by_rule = collections.defaultdict(StockMoveObj.browse)
+        applicable_rules = self.get_split_pick_applicable_rules()
+
+        for move in self:
+            move_qty = move.product_uom_qty
+            # Post assign actions could change the picking of assigned moves, so use move instead.
+            picking = move.picking_id
+
+            for rule in applicable_rules.sorted("sequence"):
+                # Iterate on the rules applicable to this picking type, attempting to reserve
+                # stock as per the configured rules strategy defines,
+                # one by one until the stock is fully reserved or until we have run out of rules.
+                if move_qty > 0:
+                    assigned_moves = super(
+                        StockMove,
+                        move.with_context(split_pick_rule=rule, split_pick_qty=move_qty),
+                    )._action_assign()
+                    # We only need to determine split characteristics if stock was assigned
+                    if assigned_moves.move_line_ids:
+                        for assigned_move in assigned_moves:
+                            # Deduct the quantity we need to consider, for any lower sequenced rules
+                            move_qty -= sum(assigned_move.move_line_ids.mapped("product_uom_qty"))
+                            new_moves = assigned_move.split_out_move_lines(assigned_move.move_line_ids)
+                            # The picking id can be cleared by split_out_move_lines, this hacks it back in.
+                            new_moves.write({"picking_id": picking.id})
+                            moves_to_split_by_rule[rule] |= new_moves
+
+        # Now we have split the lines and built the information on which rule each move is related to
+        # we can split those into backorders. Doing it this way prevents us from having picks with multiple
+        # products (moves which cannot be _merged) backorder to several separate pickings.
+        for rule, assigned_moves in moves_to_split_by_rule.items():
+            res |= assigned_moves._split_pick_for_rule(assigned_moves, rule)
+
+        if not applicable_rules:
```

Comment:
Nit: Please move this above smth like line 166 after finding applicable_rules, not to run unneccessary code if there arent applicable rules